src/main/java/org/broadinstitute/hellbender/tools/walkers/haplotypecaller/AssemblyBasedCallerUtils.java

```diff
@@ -744,6 +754,11 @@ static int constructPhaseSetMapping(final List<VariantContext> originalCalls,
                         // the phase as if the call is heterozygous and then "fix" it downstream as needed.
                         phaseSetMapping.put(call, Pair.of(uniqueCounter, PhaseGroup.PHASE_01));
                         phaseSetMapping.put(comp, Pair.of(uniqueCounter, PhaseGroup.PHASE_01));
+
+                        // if the call was on all haps but the comp isn't, we need to narrow down the set of haps we'll consider
+                        // for further phasing with the call
+                        callHaplotypesAvailableForPhasing.retainAll(haplotypesWithComp);
+
```

Comment:
I thought about doing something like this myself but got a bit scared off of making major changes to the algorithm due to its complexity -- I got some unexpected results in the integration tests when I made an initial attempt at it. However, I think that implementing the tests you suggest below and maybe the other refactorings you've suggested might make the algorithm easier to change and reason about. I'll give it another shot.

Reply:
After thinking about this some more and experimenting with it, I don't think this approach can work. A very confusing thing about this method, which has tripped me up multiple times when I've looked at issues in this code, is that its logic is not actually considering genotypes: it only considers the set of haplotypes that support alternate alleles, and doesn't get to look at the reference haplotype. Therefore it can't make the distinction between a hom var and a het which is supported by all of the alternate-allele supporting haplotypes. The comments and variable names in the method, and in the unit test, have not made this easy to understand, although the comment directly above this one obliquely refers to situations in which the set of haplotypes considered here may not be indicative of genotype. 

I've been tempted to try to modify this method to also consider the reference haplotype. The problem is that apart from the genotypes on the variant context we don't know if there was actual support in the read evidence for the ref haplotype existing. I think that considering genotypes (by peeking in to the genotypes attached to the variant contexts in the haplotype map) could be very complicated since 1) as the comment above notes, genotypes can be changed by downstream processing and 2) it looks possible to me that this method could be run in multi-sample mode, in which case you have to build a VC -> haplotype map for each sample and phase independently. Making sure this method works in multisample mode (or preventing it from being run) should probably be its own ticket, although I'd imagine it would be low priority.

Instead, I've done my best to rename variables and make the code comments more explicit about the distinction between haplotype sets and genotypes. Let me know if you have additional suggestions.